### PR TITLE
Bug Fix: Handle null images for playlist

### LINKED ISSUE #715: Spotify API error: invalid type: null
Everytime I launch Spot via Flatpak, I get a notifcation in Spot that says:
"An Error Occured. Check logs for details!"
Distro: Fedora 40 (Workstation) (Desktop)
Spot Version: 0.4.1
Logs:
```
6:47:20 PM spot: [2024-05-14T22:47:20Z ERROR spot::app::components] Spotify API error: invalid type: null, expected a sequence at line 1 column 3696
 6:47:20 PM spot: [2024-05-14T22:47:20Z ERROR spot::app::components] Spotify API error: invalid type: null, expected a sequence at line 1 column 3696
 6:47:20 PM spot: [src/api/cached_client.rs:256:17] e = Error("invalid type: null, expected a sequence", line: 1, column: 3696)
 6:47:20 PM spot: [src/api/cached_client.rs:256:17] &cache_key = "me_playlists_0_30.json"
 6:47:19 PM spot: [2024-05-14T22:47:19Z ERROR spot::app::components] Spotify API error: invalid type: null, expected a sequence at line 1 column 3696
 6:47:19 PM spot: [src/api/cached_client.rs:256:17] e = Error("invalid type: null, expected a sequence", line: 1, column: 3696)
 6:47:19 PM spot: [src/api/cached_client.rs:256:17] &cache_key = "me_playlists_0_30.json"
 6:47:17 PM spot: [src/api/cached_client.rs:256:17] e = Error("invalid type: null, expected a sequence", line: 1, column: 3696)
 6:47:17 PM spot: [src/api/cached_client.rs:256:17] &cache_key = "me_playlists_0_30.json"
 6:47:17 PM spot: GtkLabel 0x55855edf5ca0 (label) reported min height 26 and natural height 22 in measure() with for_size=-1; natural size must be >= min size
```
Stack Retrace:
```
(process:2): Adwaita-CRITICAL **: 18:55:32.911: adw_leaflet_set_visible_child: assertion 'contains_child' failed
[src/api/cached_client.rs:256:17] &cache_key = "me_playlists_0_30.json"
[src/api/cached_client.rs:256:17] e = Error("invalid type: null, expected a sequence", line: 1, column: 3696)

(spot:2): Gtk-WARNING **: 18:55:33.105: GtkLabel 0x557cce4bcca0 (label) reported min height 26 and natural height 22 in measure() with for_size=-1; natural size must be >= min size
[src/api/cached_client.rs:256:17] &cache_key = "me_playlists_0_30.json"
[src/api/cached_client.rs:256:17] e = Error("invalid type: null, expected a sequence", line: 1, column: 3696)
[2024-05-14T22:55:35Z ERROR spot::app::components] Spotify API error: invalid type: null, expected a sequence at line 1 column 3696
[src/api/cached_client.rs:256:17] &cache_key = "me_playlists_0_30.json"
[src/api/cached_client.rs:256:17] e = Error("invalid type: null, expected a sequence", line: 1, column: 3696)
[2024-05-14T22:55:35Z ERROR spot::app::components] Spotify API error: invalid type: null, expected a sequence at line 1 column 3696

### PR DESCRIPTION
**Pull request by potatoes1286**
Fixes #715, potentially #713 

If a user has a playlist with no image (due to only local songs, no songs in playlist, etc), Spotify API returns a null for image and causes Spot to fail loading any playlist changes.

This turns images to an option and, if null, returns an empty image instead. It is a quick and dirty fix, but I do not know much Rust, so I've done what I could.

Thank you in advance.

@xou816

### CODE REVIEW

**Comment by potatoes1286** [comment]
@Diegovsky @knokelmaat

**Comment by Diegovsky** [comment]
sorry for not seeing this earlier, I'm usually busy with my job so I don't receive notifications from gh's PRs


I'll look at it later today

**Comment by Diegovsky** [comment]
Also, try to target the maintenance branch as that's where current development is going

**Comment by potatoes1286** [comment]
I see. My apoligies, I will try to fix that.

File: src/api/api_models.rs [hunk 1]
```diff
@@ -197,9 +197,14 @@ pub struct PlaylistOwner {
...
 }
 
+const def_image: &'static [Image] = &[Image {url: String::new(), height: Some(640), width: Some(640)}];
+
 impl WithImages for Playlist {
     fn images(&self) -> &[Image] {
-        &self.images[..]
+        match &self.images {
+            Some(x) => &x[..],
+            None => &def_image[..],
```
Comment (Diegovsky) [inline src/api/api_models.rs]: Can't we return an empty array here?

File: src/api/api_models.rs [hunk 2]
```diff
@@ -186,7 +186,7 @@ trait WithImages {
 pub struct Playlist {
     pub id: String,
     pub name: String,
-    pub images: Vec<Image>,
+    pub images: Option<Vec<Image>>,
```
Comment (Diegovsky) [inline src/api/api_models.rs:189]: Does `#[serde(default)]` fix the issue?

**Review by Diegovsky — CHANGES_REQUESTED** [review]

**Comment by Diegovsky** [comment]
hey @potatoes1286 I added some comments but I would be happy to merge this as-is.

File: src/api/api_models.rs (hunk 2 above)
Comment (potatoes1286) [reply src/api/api_models.rs:189]: Surprisingly, no. I haven't investigated why but adding that to the images will still throw a null-errror.

File: src/api/api_models.rs (hunk 1 above)
Comment (potatoes1286) [reply src/api/api_models.rs]: No. Rust returns `returns a value referencing data owned by the current function`. I believe it has to do with garbage collection, but I do not know much rust so I just got around the issue by making a static constant outside the implementation.

**Comment by potatoes1286** [comment]
Any other concerns @Diegovsky or is it good to merge?

**Review by Diegovsky — APPROVED** [review]

**Comment by Diegovsky** [comment]
@potatoes1286 please run `cargo fmt`

**Comment by Diegovsky** [comment]
I just want to make all the tests pass before merge. Looks fine by me.

I apologise for the long review process. Thank you for your contribution :)

**Comment by potatoes1286** [comment]
@Diegovsky Formatted now 👍

Side note: renamed def_image -> EMPTY_IMAGE to match rust naming / more reflective naming

**Review by Diegovsky — APPROVED** [review]

**Comment by Diegovsky** [comment]
Thank you for your work!

**Pull request merged**